libs/h3/src/main/java/org/elasticsearch/h3/HexRing.java

```diff
@@ -581,52 +599,116 @@ public static long[] hexRing(long origin) {
+     * Returns whether or not the provided H3Indexes are neighbors.
+     * @param origin The origin H3 index.
+     * @param destination The destination H3 index.
+     * @return true if the indexes are neighbors, false otherwise
+     */
+    public static boolean areNeighbours(long origin, long destination) {
+        // Make sure they're hexagon indexes
+        if (H3Index.H3_get_mode(origin) != Constants.H3_CELL_MODE) {
+            throw new IllegalArgumentException("Invalid cell: " + origin);
+        }
+
+        if (H3Index.H3_get_mode(destination) != Constants.H3_CELL_MODE) {
+            throw new IllegalArgumentException("Invalid cell: " + destination);
+        }
+
+        // Hexagons cannot be neighbors with themselves
+        if (origin == destination) {
+            return false;
+        }
+
+        final int resolution = H3Index.H3_get_resolution(origin);
+        // Only hexagons in the same resolution can be neighbors
+        if (resolution != H3Index.H3_get_resolution(destination)) {
+            return false;
+        }
+
+        // H3 Indexes that share the same parent are very likely to be neighbors
+        // Child 0 is neighbor with all of its parent's 'offspring', the other
+        // children are neighbors with 3 of the 7 children. So a simple comparison
+        // of origin and destination parents and then a lookup table of the children
+        // is a super-cheap way to possibly determine they are neighbors.
+        if (resolution > 1) {
+            long originParent = H3.h3ToParent(origin);
+            long destinationParent = H3.h3ToParent(destination);
+            if (originParent == destinationParent) {
+                int originResDigit = H3Index.H3_get_index_digit(origin, resolution);
+                int destinationResDigit = H3Index.H3_get_index_digit(destination, resolution);
+                if (originResDigit == CoordIJK.Direction.CENTER_DIGIT.digit()
+                    || destinationResDigit == CoordIJK.Direction.CENTER_DIGIT.digit()) {
+                    return true;
+                }
+                if (originResDigit >= CoordIJK.Direction.INVALID_DIGIT.digit()) {
+                    // Prevent indexing off the end of the array below
+                    throw new IllegalArgumentException("");
+                }
+                if ((originResDigit == CoordIJK.Direction.K_AXES_DIGIT.digit()
+                    || destinationResDigit == CoordIJK.Direction.K_AXES_DIGIT.digit()) && H3.isPentagon(originParent)) {
+                    // If these are invalid cells, fail rather than incorrectly
+                    // reporting neighbors. For pentagon cells that are actually
+                    // neighbors across the deleted subsequence, they will fail the
+                    // optimized check below, but they will be accepted by the
+                    // gridDisk check below that.
+                    throw new IllegalArgumentException("Undefined error checking for neighbors");
```

Comment:
I have to admit the errors thrown in the code are very uninformative. Most of them are empty strings, and this one is almost as uninformative. I guess these all come from the original H3 code.

Reply:
I must say I don't know how to improve it. Please feel free to do it.